File: internal/manx/parser.go
package manx

import "strings"

type Segment struct {
	Raw        string
	Sudo       bool
	Command    string
	Args       []string
	Options    []string
	RawOptions []string
	Operands   []string
	Redirects  []string
}

func (s Segment) HasOption(opt string) bool {
	for _, o := range s.Options {
		if o == opt {
			return true
		}
	}
	return false
}

type ParsedCommand struct {
	Raw                    string
	Segments               []Segment
	HasPipe                bool
	HasCommandSubstitution bool
	HasGlob                bool
	PipeToShell            bool
}

func (p ParsedCommand) Commands() []string {
	var out []string
	for _, s := range p.Segments {
		if s.Command != "" {
			out = append(out, s.Command)
		}
	}
	return out
}

func Parse(line string) ParsedCommand {
	line = strings.TrimSpace(line)
	pc := ParsedCommand{
		Raw:                    line,
		HasCommandSubstitution: strings.Contains(line, "$(") || strings.Contains(line, "`"),
		HasGlob:                strings.ContainsAny(line, "*?"),
	}
	parts := splitPipes(line)
	pc.HasPipe = len(parts) > 1
	for _, part := range parts {
		pc.Segments = append(pc.Segments, ParseSegment(part))
	}
	if pc.HasPipe {
		hasDownloader := false
		hasShell := false
		for _, cmd := range pc.Commands() {
			switch cmd {
			case "curl", "wget", "fetch":
				hasDownloader = true
			case "bash", "sh", "zsh", "dash", "ksh", "fish":
				hasShell = true
			}
		}
		pc.PipeToShell = hasDownloader && hasShell
	}
	return pc
}

func ParseSegment(raw string) Segment {
	seg := Segment{Raw: strings.TrimSpace(raw)}
	tokens := shellFields(raw)
	if len(tokens) == 0 {
		return seg
	}

	idx := 0
	for idx < len(tokens) && (tokens[idx] == "sudo" || tokens[idx] == "doas") {
		seg.Sudo = true
		idx++
		for idx < len(tokens) && strings.HasPrefix(tokens[idx], "-") {
			opt := tokens[idx]
			idx++
			if opt == "-u" || opt == "--user" || opt == "-g" || opt == "--group" {
				idx++
			}
		}
	}
	tokens = tokens[idx:]
	if len(tokens) == 0 {
		return seg
	}

	seg.Command = tokens[0]
	expectRedirectTarget := false
	for _, tok := range tokens[1:] {
		if expectRedirectTarget {
			seg.Redirects = append(seg.Redirects, tok)
			expectRedirectTarget = false
			continue
		}
		if isRedirect(tok) {
			seg.Redirects = append(seg.Redirects, tok)
			expectRedirectTarget = true
			continue
		}
		seg.Args = append(seg.Args, tok)
		switch {
		case strings.HasPrefix(tok, "--"):
			base := strings.SplitN(tok, "=", 2)[0]
			seg.Options = append(seg.Options, base)
			seg.RawOptions = append(seg.RawOptions, base)
		case strings.HasPrefix(tok, "-") && len(tok) > 1:
			seg.RawOptions = append(seg.RawOptions, tok)
			seg.Options = append(seg.Options, explodeOptions(tok)...)
		default:
			seg.Operands = append(seg.Operands, tok)
		}
	}
	return seg
}

func splitPipes(line string) []string {
	var parts []string
	var b strings.Builder
	var quote rune
	runes := []rune(line)
	for i := 0; i < len(runes); i++ {
		ch := runes[i]
		if quote != 0 {
			b.WriteRune(ch)
			if ch == quote {
				quote = 0
			}
			continue
		}
		if ch == '\'' || ch == '"' {
			quote = ch
			b.WriteRune(ch)
			continue
		}
		if ch == '|' {
			if i+1 < len(runes) && runes[i+1] == '|' {
				b.WriteString("||")
				i++
				continue
			}
			part := strings.TrimSpace(b.String())
			if part != "" {
				parts = append(parts, part)
			}
			b.Reset()
			continue
		}
		b.WriteRune(ch)
	}
	if part := strings.TrimSpace(b.String()); part != "" {
		parts = append(parts, part)
	}
	return parts
}

func shellFields(s string) []string {
	var fields []string
	var b strings.Builder
	var quote rune
	escaped := false
	have := false
	for _, ch := range s {
		if escaped {
			b.WriteRune(ch)
			have = true
			escaped = false
			continue
		}
		if ch == '\\' && quote != '\'' {
			escaped = true
			have = true
			continue
		}
		if quote != 0 {
			if ch == quote {
				quote = 0
			} else {
				b.WriteRune(ch)
			}
			have = true
			continue
		}
		if ch == '\'' || ch == '"' {
			quote = ch
			have = true
			continue
		}
		if ch == ' ' || ch == '\t' || ch == '\n' {
			if have {
				fields = append(fields, b.String())
				b.Reset()
				have = false
			}
			continue
		}
		b.WriteRune(ch)
		have = true
	}
	if escaped {
		b.WriteRune('\\')
	}
	if have {
		fields = append(fields, b.String())
	}
	return fields
}

func explodeOptions(token string) []string {
	if strings.HasPrefix(token, "--") || token == "-" || !strings.HasPrefix(token, "-") {
		return nil
	}
	var opts []string
	for _, ch := range token[1:] {
		if (ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'Z') {
			opts = append(opts, "-"+string(ch))
		}
	}
	return opts
}

func isRedirect(tok string) bool {
	if tok == ">" || tok == ">>" || tok == "<" || tok == "&>" {
		return true
	}
	if strings.HasSuffix(tok, ">") || strings.HasSuffix(tok, ">>") {
		prefix := strings.TrimRight(tok, ">")
		if prefix == "" {
			return true
		}
		for _, ch := range prefix {
			if ch < '0' || ch > '9' {
				return false
			}
		}
		return true
	}
	if strings.Contains(tok, ">&") {
		return true
	}
	return false
}

File: src/manx/parser.py
"""命令解析层：把一行命令拆成可分析的结构。

MVP 采用轻量解析（不接完整 Bash AST），识别：
管道 | 、重定向 > >> 、sudo、短参数组合、长参数、路径、通配符、命令替换 $()。
"""

from __future__ import annotations

import re
import shlex
from dataclasses import dataclass, field
from typing import List, Optional


@dataclass
class Segment:
    """一条由管道分隔的子命令。"""

    raw: str
    sudo: bool = False
    command: str = ""
    args: List[str] = field(default_factory=list)
    options: List[str] = field(default_factory=list)       # 展开后的单个选项：-x -z -f --recursive（供风险引擎按字母判断）
    raw_options: List[str] = field(default_factory=list)    # 原样选项 token：-xzf -name --recursive（供结构展示）
    operands: List[str] = field(default_factory=list)       # 非选项参数（路径/值）
    redirects: List[str] = field(default_factory=list)

    def has_option(self, opt: str) -> bool:
        return opt in self.options


@dataclass
class ParsedCommand:
    raw: str
    segments: List[Segment] = field(default_factory=list)
    has_pipe: bool = False
    has_command_substitution: bool = False
    has_glob: bool = False
    pipe_to_shell: bool = False   # curl ... | bash 之类

    @property
    def commands(self) -> List[str]:
        return [s.command for s in self.segments if s.command]


_REDIRECT_RE = re.compile(r"^\d*>>?$|^<$|^&>$|^\d*>&\d*$")
_SHELL_RUNNERS = {"bash", "sh", "zsh", "dash", "ksh", "fish"}


def _split_pipes(line: str) -> List[str]:
    """按未被引号包裹的 | 切分（忽略 ||）。"""
    parts: List[str] = []
    buf = ""
    quote = ""
    i = 0
    while i < len(line):
        ch = line[i]
        if quote:
            buf += ch
            if ch == quote:
                quote = ""
        elif ch in "\"'":
            quote = ch
            buf += ch
        elif ch == "|":
            if i + 1 < len(line) and line[i + 1] == "|":  # || 逻辑或，MVP 不深入
                buf += "||"
                i += 2
                continue
            parts.append(buf)
            buf = ""
        else:
            buf += ch
        i += 1
    if buf.strip():
        parts.append(buf)
    return [p.strip() for p in parts if p.strip()]


def _explode_options(token: str) -> List[str]:
    """把 -xzf 展开成 -x -z -f；长参数保持整体。"""
    if token.startswith("--"):
        return [token]
    if token.startswith("-") and len(token) > 1 and token != "-":
        body = token[1:]
        # 把组合短参拆成单个选项字母，例如 -xzf -> -x -z -f
        return [f"-{c}" for c in body if c.isalpha()]
    return []


def parse_segment(raw: str) -> Segment:
    seg = Segment(raw=raw.strip())
    try:
        tokens = shlex.split(raw, comments=False)
    except ValueError:
        tokens = raw.split()

    if not tokens:
        return seg

    # sudo（可能带 -u 等，MVP 简单处理：跳过 sudo 及其纯选项）
    idx = 0
    while idx < len(tokens) and tokens[idx] in ("sudo", "doas"):
        seg.sudo = True
        idx += 1
        # 跳过 sudo 自身的选项（-u user, -E 等）
        while idx < len(tokens) and tokens[idx].startswith("-"):
            opt = tokens[idx]
            idx += 1
            if opt in ("-u", "--user", "-g", "--group"):
                idx += 1  # 跳过其值
    tokens = tokens[idx:]
    if not tokens:
        return seg

    seg.command = tokens[0]
    rest = tokens[1:]

    expect_redirect_target = False
    for tok in rest:
        if expect_redirect_target:
            seg.redirects.append(tok)
            expect_redirect_target = False
            continue
        if _REDIRECT_RE.match(tok):
            seg.redirects.append(tok)
            expect_redirect_target = True
            continue
        seg.args.append(tok)
        if tok.startswith("--"):
            base = tok.split("=", 1)[0]
            seg.options.append(base)
            seg.raw_options.append(base)
        elif tok.startswith("-") and len(tok) > 1:
            # 既保留原 token（-xzf / -name），又展开成单字母供风险引擎判断
            seg.raw_options.append(tok)
            seg.options.extend(_explode_options(tok))
        else:
            seg.operands.append(tok)
    return seg


def parse(line: str) -> ParsedCommand:
    line = line.strip()
    pc = ParsedCommand(raw=line)
    pc.has_command_substitution = bool(re.search(r"\$\(|`", line))
    pc.has_glob = bool(re.search(r"(^|\s)[^\s]*[*?]", line)) or "*" in line

    parts = _split_pipes(line)
    pc.has_pipe = len(parts) > 1
    for p in parts:
        pc.segments.append(parse_segment(p))

    # 管道执行远程脚本：上游是 curl/wget，下游是 shell
    if pc.has_pipe:
        cmds = pc.commands
        if any(c in ("curl", "wget", "fetch") for c in cmds) and any(
            c in _SHELL_RUNNERS for c in cmds
        ):
            pc.pipe_to_shell = True

    return pc
